File: stacks/fastapi/base/code/middleware.tpl.py
"""
File: middleware.tpl.py
Purpose: Custom middleware for logging, monitoring, and request processing
Generated for: {{PROJECT_NAME}}
Tier: base
Stack: fastapi
Category: middleware
"""

from fastapi import Request, Response
from fastapi.responses import JSONResponse
from starlette.middleware.base import BaseHTTPMiddleware
from typing import Callable
import time
import logging
import uuid

logger = logging.getLogger(__name__)


# ============================================================================
# Request ID Middleware
# ============================================================================

class RequestIDMiddleware(BaseHTTPMiddleware):
    """
    Adds a unique request ID to each request for tracing.
    """
    
    async def dispatch(
        self, request: Request, call_next: Callable
    ) -> Response:
        # Generate unique request ID
        request_id = str(uuid.uuid4())
        
        # Add to request state
        request.state.request_id = request_id
        
        # Process request
        response = await call_next(request)
        
        # Add request ID to response headers
        response.headers["X-Request-ID"] = request_id
        
        return response


# ============================================================================
# Logging Middleware
# ============================================================================

class LoggingMiddleware(BaseHTTPMiddleware):
    """
    Logs all requests and responses with timing information.
    """
    
    async def dispatch(
        self, request: Request, call_next: Callable
    ) -> Response:
        # Start timer
        start_time = time.time()
        
        # Get request ID if available
        request_id = getattr(request.state, "request_id", "N/A")
        
        # Log request
        logger.info(
            f"Request started",
            extra={
                "request_id": request_id,
                "method": request.method,
                "path": request.url.path,
                "client_host": request.client.host if request.client else None,
            }
        )
        
        # Process request
        response = await call_next(request)
        
        # Calculate duration
        duration = time.time() - start_time
        
        # Log response
        logger.info(
            f"Request completed",
            extra={
                "request_id": request_id,
                "method": request.method,
                "path": request.url.path,
                "status_code": response.status_code,
                "duration_ms": round(duration * 1000, 2),
            }
        )
        
        return response


# ============================================================================
# Rate Limiting Middleware
# ============================================================================

from collections import defaultdict
from datetime import datetime, timedelta

class RateLimitMiddleware(BaseHTTPMiddleware):
    """
    Simple in-memory rate limiting middleware.
    For production, use Redis-based rate limiting.
    """
    
    def __init__(self, app, calls: int = 100, period: int = 60):
        super().__init__(app)
        self.calls = calls  # Max calls per period
        self.period = period  # Period in seconds
        self.clients = defaultdict(list)
    
    async def dispatch(
        self, request: Request, call_next: Callable
    ) -> Response:
        # Get client identifier
        client_id = request.client.host if request.client else "unknown"
        
        # Clean old timestamps
        now = datetime.utcnow()
        cutoff = now - timedelta(seconds=self.period)
        self.clients[client_id] = [
            ts for ts in self.clients[client_id]
            if ts > cutoff
        ]
        
        # Check rate limit
        if len(self.clients[client_id]) >= self.calls:
            return JSONResponse(
                status_code=429,
                content={
                    "detail": "Rate limit exceeded",
                    "retry_after": self.period
                },
                headers={"Retry-After": str(self.period)}
            )
        
        # Record this request
        self.clients[client_id].append(now)
        
        # Process request
        response = await call_next(request)
        
        # Add rate limit headers
        remaining = self.calls - len(self.clients[client_id])
        response.headers["X-RateLimit-Limit"] = str(self.calls)
        response.headers["X-RateLimit-Remaining"] = str(remaining)
        response.headers["X-RateLimit-Reset"] = str(int((now + timedelta(seconds=self.period)).timestamp()))
        
        return response


# ============================================================================
# Error Handling Middleware
# ============================================================================

class ErrorHandlingMiddleware(BaseHTTPMiddleware):
    """
    Catches and formats unhandled exceptions.
    """
    
    async def dispatch(
        self, request: Request, call_next: Callable
    ) -> Response:
        try:
            response = await call_next(request)
            return response
        except Exception as e:
            request_id = getattr(request.state, "request_id", "N/A")
            
            logger.error(
                f"Unhandled exception",
                extra={
                    "request_id": request_id,
                    "error": str(e),
                    "error_type": type(e).__name__,
                },
                exc_info=True
            )
            
            return JSONResponse(
                status_code=500,
                content={
                    "detail": "Internal server error",
                    "request_id": request_id,
                },
                headers={"X-Request-ID": request_id}
            )


# ============================================================================
# CORS Headers Middleware (if not using CORSMiddleware)
# ============================================================================

class CustomCORSMiddleware(BaseHTTPMiddleware):
    """
    Custom CORS middleware with more control.
    For most cases, use FastAPI's built-in CORSMiddleware.
    """
    
    def __init__(self, app, allowed_origins: list = None):
        super().__init__(app)
        self.allowed_origins = allowed_origins or ["*"]
    
    async def dispatch(
        self, request: Request, call_next: Callable
    ) -> Response:
        # Handle preflight requests
        if request.method == "OPTIONS":
            response = Response()
            response.headers["Access-Control-Allow-Origin"] = "*"
            response.headers["Access-Control-Allow-Methods"] = "GET, POST, PUT, DELETE, OPTIONS"
            response.headers["Access-Control-Allow-Headers"] = "Content-Type, Authorization"
            return response
        
        # Process request
        response = await call_next(request)
        
        # Add CORS headers
        origin = request.headers.get("origin")
        if origin in self.allowed_origins or "*" in self.allowed_origins:
            response.headers["Access-Control-Allow-Origin"] = origin or "*"
            response.headers["Access-Control-Allow-Credentials"] = "true"
        
        return response


# ============================================================================
# Security Headers Middleware
# ============================================================================

class SecurityHeadersMiddleware(BaseHTTPMiddleware):
    """
    Adds security headers to all responses.
    """
    
    async def dispatch(
        self, request: Request, call_next: Callable
    ) -> Response:
        response = await call_next(request)
        
        # Add security headers
        response.headers["X-Content-Type-Options"] = "nosniff"
        response.headers["X-Frame-Options"] = "DENY"
        response.headers["X-XSS-Protection"] = "1; mode=block"
        response.headers["Strict-Transport-Security"] = "max-age=31536000; includeSubDomains"
        response.headers["Content-Security-Policy"] = "default-src 'self'"
        
        return response

File: _complete_archive/_archive_2026-02-26_18-50-32/stacks/node/base/code/middleware.tpl.js
/*
File: middleware.tpl.js
Purpose: Express middleware for common patterns
Generated for: {{PROJECT_NAME}}
*/

const { v4: uuidv4 } = require('uuid');

/**
 * Request ID middleware - adds unique ID to each request
 */
function requestId() {
    return (req, res, next) => {
        req.id = req.headers['x-request-id'] || uuidv4();
        res.setHeader('X-Request-ID', req.id);
        next();
    };
}

/**
 * Request logging middleware
 */
function requestLogger(options = {}) {
    const { logger = console } = options;

    return (req, res, next) => {
        const start = Date.now();

        res.on('finish', () => {
            const duration = Date.now() - start;
            logger.info({
                requestId: req.id,
                method: req.method,
                path: req.path,
                statusCode: res.statusCode,
                duration: `${duration}ms`,
            });
        });

        next();
    };
}

/**
 * Rate limiting middleware (in-memory, use Redis in production)
 */
function rateLimit(options = {}) {
    const {
        windowMs = 60000, // 1 minute
        max = 100, // requests per window
        message = 'Too many requests, please try again later',
    } = options;

    const requests = new Map();

    // Cleanup old entries every minute
    setInterval(() => {
        const now = Date.now();
        for (const [key, data] of requests.entries()) {
            if (now - data.firstRequest > windowMs) {
                requests.delete(key);
            }
        }
    }, windowMs);

    return (req, res, next) => {
        const key = req.ip;
        const now = Date.now();

        if (!requests.has(key)) {
            requests.set(key, { count: 1, firstRequest: now });
        } else {
            const data = requests.get(key);
            if (now - data.firstRequest > windowMs) {
                requests.set(key, { count: 1, firstRequest: now });
            } else {
                data.count++;
            }
        }

        const data = requests.get(key);
        const remaining = Math.max(0, max - data.count);

        res.setHeader('X-RateLimit-Limit', max);
        res.setHeader('X-RateLimit-Remaining', remaining);
        res.setHeader('X-RateLimit-Reset', Math.ceil((data.firstRequest + windowMs) / 1000));

        if (data.count > max) {
            res.status(429).json({ error: message });
            return;
        }

        next();
    };
}

/**
 * Error handling middleware
 */
function errorHandler(options = {}) {
    const { logger = console, showStack = process.env.NODE_ENV !== 'production' } = options;

    return (err, req, res, next) => {
        logger.error({
            requestId: req.id,
            error: err.message,
            stack: err.stack,
        });

        const statusCode = err.statusCode || err.status || 500;
        const response = {
            error: {
                message: err.message || 'Internal Server Error',
                code: err.code || 'INTERNAL_ERROR',
            },
        };

        if (showStack) {
            response.error.stack = err.stack;
        }

        res.status(statusCode).json(response);
    };
}

/**
 * Not found handler
 */
function notFound() {
    return (req, res) => {
        res.status(404).json({
            error: {
                message: 'Resource not found',
                code: 'NOT_FOUND',
                path: req.path,
            },
        });
    };
}

/**
 * Security headers middleware
 */
function securityHeaders() {
    return (req, res, next) => {
        res.setHeader('X-Content-Type-Options', 'nosniff');
        res.setHeader('X-Frame-Options', 'DENY');
        res.setHeader('X-XSS-Protection', '1; mode=block');
        res.setHeader('Strict-Transport-Security', 'max-age=31536000; includeSubDomains');
        res.removeHeader('X-Powered-By');
        next();
    };
}

/**
 * Response time header middleware
 */
function responseTime() {
    return (req, res, next) => {
        const start = process.hrtime();

        res.on('finish', () => {
            const diff = process.hrtime(start);
            const time = (diff[0] * 1e3 + diff[1] * 1e-6).toFixed(2);
            res.setHeader('X-Response-Time', `${time}ms`);
        });

        next();
    };
}

/**
 * Setup all middleware
 */
function setupMiddleware(app) {
    app.use(requestId());
    app.use(responseTime());
    app.use(requestLogger());
    app.use(securityHeaders());
    app.use(rateLimit());
}

module.exports = {
    requestId,
    requestLogger,
    rateLimit,
    errorHandler,
    notFound,
    securityHeaders,
    responseTime,
    setupMiddleware,
};

// Usage:
// const express = require('express');
// const { setupMiddleware, errorHandler, notFound } = require('./middleware');
//
// const app = express();
// setupMiddleware(app);
// // ... routes ...
// app.use(notFound());
// app.use(errorHandler());
